### PR TITLE
Add spawnpoint display

### PR DESCRIPTION
**Pull request by Xcelled**
Shows spawnpoints on the map, based on https://www.reddit.com/r/pokemongodev/comments/4wxteh/i_implemented_tbterras_spawntracker_into/

## Description
While this does not change scanning behavior, it does display all known spawnpoints on the map for informational purposes. Spawnpoints are highlighted as a 5m circle and can be toggled on/off in options, defaulting to off.

## Motivation and Context
Spawnpoint based scanning is apparently the Next Big Thing (tm), and while I didn't want to implement the whole scan stuff, I thought it'd be informative to see where the spawnpoints actually are, if for no other reason than to laugh at how inefficient beehive is.

## How Has This Been Tested?
Spot-checked on Win10 with Chrome, and Android 6.0.1 with Chrome.
Spawnpoint output was (roughly) compared to the output generated from the SQLite query on the Reddit page.

## Screenshots (if appropriate):
Pokemon Off:
image

Pokemon On:
image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation. (Probably, I don't know if the UI options are documented)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by bbbenji** [comment]
Please provide a way to to export spawnpoints.

**Comment by Xcelled** [comment]
@bbbenji That's probably a little out of the scope of this pull request (more of a tool), but the code I added in models.py could easily be reused for such a purpose.

**Comment by bbbenji** [comment]
@Xcelled Got it working, thanks.

**Comment by invisiblek** [comment]
I'm not sure how I feel about this. We're already having performance issues with scanned locations: https://github.com/PokemonGoMap/PokemonGo-Map/pull/386

Wouldn't this particular feature introduce even more performance issues than https://github.com/PokemonGoMap/PokemonGo-Map/pull/386 ?

**Comment by Infernoman** [comment]
This should reduce the queries on the servers both ours and niantic's by only scanning certain spawn points at the intervals they are supposed to spawn, rather than continuously scanning all area's see > https://www.reddit.com/r/pokemongodev/comments/4wxteh/i_implemented_tbterras_spawntracker_into

**Comment by invisiblek** [comment]
That patch won't make it into our code base (as-is at least). The idea of skipping scan locations is something I can't agree with.

Now, if it could be improved to be smart enough to prioritize locations and scan those more often, we could talk.

Regardless, its not a fix to what I'm referring to.

**Comment by Xcelled** [comment]
This doesn't skip locations, or impact the scanning at all. It's strictly informational only.

There's no caching of the spawnpoints, primarily because I didn't want to make too many changes to the underlying architecture, since this does not actually implement spawnpoint scanning. This could be something to implement if performance is an issue.

However, I don't think that'll be needed. This, inlike scanned locations, isn't meant to be turned on constantly/over a wide area. That's not particularly useful, as the points are pretty small and get hard to see at too far out of a zoom level. I also noticed some possible inefficiency in the js that could be fixed.

Regardless, if there are performance concerns, conducting profiling is the way to go. I can run some tests.

**Comment by Xcelled** [comment]
I Instrumented UpdateMap in the JS and did a 7 step scan of the Santa Monica pier without the spawnpoint code and timed approximately 1,000 invocations. Then, I added the spawnpoints calls and repeated the process. There were a LOT of spawnpoints, as you can see in this picture:

image

Without the spawnpoints, UpdateMap took an average of 373 ms to run. With it, 378.

Unless there's another method I should look at, the cpu usage doesn't seem terribly affected. Of course, if you zoom out enough, eventually you'll get bogged down by spawn points, but that's true of any marker on the map.

I'd also like to note that at least four people have directly told me this is useful to them, and one requested I send them screenshots of the spawnpoints around them.

**Comment by cpius** [comment]
+1

Let's get the next big thing already!

**Comment by Goldvivi1** [comment]
@xcelled looks cool and useful when zoomed in, when zooming out in the UI they are useless. Maybe using the Viewport SW and NE points to calculate if the spawn points are to be displayed or not.

**Comment by Xcelled** [comment]
@Goldvivi1 that's similar to what's being suggested in #386 for scanned points. My concern with that would be screen size, ie a smartphone screen can only display a small geographical location at a zoom where spawnpoints are useful, but two 24inch monitors could display a much larger one.

I think that for now, it should just be up to the user to turn them on/off as desired. If/when #386 is resolved, hopefully the same fix could be applied here.

**Comment by schadis** [comment]
pls and now make it possible in config that it only scans the whole map like 2 times in a hour for new spawn points and the rest of the time only spawn points shall be scanned +1

**Comment by Xcelled** [comment]
@schadis see #179

**Comment by Xcelled** [comment]
I just wanted to note that, even though the reddit SQL's group by has changed, this pull request does not need to be updated. The reddit group by's purpose is to select unique combinations of spawnpoint coords and times.

This code doesn't calculate or care about the time, which eliminates the need to group on it. spawnpoint_id maps a 1-to-1 to a unique lat/long pair, so grouping on the spawnpoint_id effectively does the same as grouping on the lat/lon pairs.

**TL;DR:** No changes needed.

**Comment by chuyskywalker** [comment]
Awesome. A good step in the direction of being able to change up towards waaaay more efficient tracking.

**Comment by rjwestman** [comment]
I sometimes get spawnpoints, but not for every pokemon on the map.

**Comment by jcat00uk** [comment]
is it feasible to have more info on the spawnpoint displayed, say when you click on a spawnpoint, it brings up the hourly spawntime for that point and maybe the last spawned pokemon at that location

cheers
John
